azurelinuxagent/ga/exthandlers.py

```diff
@@ -784,17 +786,16 @@ def get_installed_version(self):
             version_from_path = FlexibleVersion(path[separator + 1:])
             state_path = os.path.join(path, 'config', 'HandlerState')
 
-            if not os.path.exists(state_path) or \
-                    fileutil.read_file(state_path) == \
-                    ExtHandlerState.NotInstalled:
-                logger.verbose("Ignoring version of uninstalled extension: "
+            if not os.path.exists(state_path) or fileutil.read_file(state_path) == ExtHandlerState.NotInstalled \
+                    or fileutil.read_file(state_path) == ExtHandlerState.FailedUpgrade:
```

Comment:
why is the extra check needed?

Reply:
As the name of the method suggests, the purpose is to find the latest installed extension version. Without the added check to ignore `FailedUpgrade` states, when a second goal state comes in (so, when the old version is disabled and the new one is failed), this method would be called through `handle_enable` and would actually retrieve the new version as the `old_exthandler` simply because it's state is **not** NotInstalled, which is incorrect.

Reply:
yes, that what it looks like, but is that happening with the current code?

Reply:
(i want to double-check we are not inadvertently changing some other behavior)

I don't remember seeing a case like that, though most of the upgrade failures we see are on disable, and not on update.

Reply:
In the current code (pre-this PR), we'd retrieve the new (failed) extension as the `old_exthandler_i` in [this line](https://github.com/Azure/WALinuxAgent/blob/develop/azurelinuxagent/ga/exthandlers.py#L482), but, because we'd jump straight to [this line](https://github.com/Azure/WALinuxAgent/blob/develop/azurelinuxagent/ga/exthandlers.py#L497), `old_exthandler_i` wasn't being used anywhere. So basically one bug was hiding the other.

Reply:
ok